File: built-in plugins dsp/mb-wt-cinematic.cpp
/**
 * MB Cinematic WT
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Epic cinematic wavetable
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WT_CINEMATIC_H
#define MB_WT_CINEMATIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWtCinematic : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-wt-cinematic";
    static constexpr const char* PLUGIN_NAME    = "MB Cinematic WT";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float position = 0.3f;  // range [0, 1]
    float epic = 0.8f;  // range [0, 1]
    float volume = 0.75f;  // range [0, 1]
    };

    MbWtCinematic() = default;
    ~MbWtCinematic() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.position = std::clamp(params.position, 0f, 1f);
        params.epic = std::clamp(params.epic, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Cinematic WT
        return input;
    }
};

#endif // MB_WT_CINEMATIC_H

File: built-in plugins dsp/mb-wt-cinematic.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbWtCinematicPlugin: PluginDefinition = {
  id: "mb-wt-cinematic",
  slug: "mb-wt-cinematic",
  name: "MB Cinematic WT",
  category: "instrument",
  type: "wavetable",
  version: "1.0.0",
  description: "Epic cinematic wavetable",
  author: "Max Booster",
  grade: "A",
  oscillators: [{ type: "sawtooth", detune: 0, gain: 0.6 }],
  envelope: { attack: 1.5, decay: 1.0, sustain: 0.85, release: 2.5 },
  parameters: [
    {
      id: "position",
      name: "Wave Position",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "epic",
      name: "Epic",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.75,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { position: 0.3, epic: 0.8, volume: 0.75 },
};

export default MbWtCinematicPlugin;
